scipy/signal/_splinemodule.cc

```diff
@@ -0,0 +1,513 @@
+                                "    Parameters\n"
+                                "    ----------\n"
+                                "    input : ndarray\n"
+                                "        The input signal. Must be a rank-2 array.\n"
+                                "    hrow : ndarray\n"
+                                "        A rank-1 array defining the row direction of the filter.\n"
+                                "        Must be odd-length\n"
+                                "    hcol : ndarray\n"
+                                "        A rank-1 array defining the column direction of the filter.\n"
+                                "        Must be odd-length\n"
+                                "\n"
+                                "    Returns\n"
+                                "    -------\n"
+                                "    output : ndarray\n"
+                                "        The filtered signal.\n"
+                                "\n"
+                                "    Examples\n"
+                                "    --------\n"
+                                "    Examples are given :ref:`in the tutorial <tutorial-signal-bsplines>`.\n"
+                                "\n";
+
+static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args) {
+    PyObject *image = NULL, *hrow = NULL, *hcol = NULL;
+    PyArrayObject *a_image = NULL, *a_hrow = NULL, *a_hcol = NULL, *out = NULL;
+    int thetype, M, N, ret;
+    npy_intp outstrides[2], instrides[2];
+
+    if (!PyArg_ParseTuple(args, "OOO", &image, &hrow, &hcol))
+        return NULL;
+
+    thetype = PyArray_ObjectType(image, NPY_FLOAT);
+    thetype = PyArray_MIN(thetype, NPY_CDOUBLE);
+    a_image = (PyArrayObject *) PyArray_FromObject(image, thetype, 2, 2);
+    if (a_image == NULL)
+        goto fail;
+
+    a_hrow = (PyArrayObject *) PyArray_ContiguousFromObject(hrow, thetype, 1, 1);
+    if (a_hrow == NULL)
+        goto fail;
+
+    a_hcol = (PyArrayObject *) PyArray_ContiguousFromObject(hcol, thetype, 1, 1);
+    if (a_hcol == NULL)
+        goto fail;
+
+    out = (PyArrayObject *) PyArray_SimpleNew(2, PyArray_DIMS(a_image), thetype);
+    if (out == NULL)
+        goto fail;
+
+    M = PyArray_DIMS(a_image)[0];
+    N = PyArray_DIMS(a_image)[1];
+
+    convert_strides(PyArray_STRIDES(a_image), instrides, PyArray_ITEMSIZE(a_image), 2);
+    outstrides[0] = N;
+    outstrides[1] = 1;
+
+    if (PyArray_DIMS(a_hrow)[0] % 2 != 1 || PyArray_DIMS(a_hcol)[0] % 2 != 1) {
+        PYERR("hrow and hcol must be odd length");
+    }
+
+    switch (thetype) {
+    case NPY_FLOAT:
+
+        ret = _separable_2Dconvolve_mirror((float *) PyArray_DATA(a_image), (float *) PyArray_DATA(out), M, N,
+                                           (float *) PyArray_DATA(a_hrow), (float *) PyArray_DATA(a_hcol),
+                                           PyArray_DIMS(a_hrow)[0], PyArray_DIMS(a_hcol)[0], instrides, outstrides);
+        break;
```

Comment:
still wondering if we can get rid of this switch and use `thetype` (suitably mapped to C++) during casting.

Reply:
Not sure how? Am open to suggestions
(so far I was trying to minimize non-trivial changes to the logic, hence the switch just lives on from the original)

Reply:
I played around with this, and it doesn't work (or my C++ foo isn't good enough).